### PR TITLE
Metadata: Preserve FK property name when setting Principal Key

### LINKED ISSUE #13300: Order of entity mappings matters when dealing with shadow properties
Our `User` entity has a One-To-many relationship with another entity called `Profile`. 

You may find the mappings we are using below:
```C#
        builder.Entity<User>(m =>
            {
                m.ToTable("User", "user");
                m.HasKey(x => x.Id);
                m.Property(x => x.Id)
                    .HasColumnName("UserId");
                m.Property("_email")
                    .HasColumnName("Email");
                m.Property(x => x.AdditionalEmail)
                    .HasConversion(ValueConverters.EmailConverter);
              
               << Relationship definition
                m.HasMany<Profile>("_profiles")
                    .WithOne()
                    .HasForeignKey("Email")
                    .HasPrincipalKey("_email");

                m.HasMany<UserRole>("_roles")
                    .WithOne()
                    .HasForeignKey(x => x.Id)
                    .HasPrincipalKey(x => x.Id);

                m.Ignore(x => x.Email);
                m.Ignore(x => x.Profiles);
                m.Ignore(x => x.Roles);
            });

            builder.Entity<Profile>(m =>
            {
                m.ToTable("Profile", "user");
                m.HasKey(x => x.Id);
                m.Property(x => x.Id)
                    .HasColumnName("ProfileId");
                m.Property(x => x.DirectLine)
                    .HasColumnName("PhoneNumber")
                    .HasConversion(ValueConverters.PhoneNumberConverter);
                m.Property(x => x.MobileNumber)
                    .HasConversion(ValueConverters.PhoneNumberConverter);
                m.Property(x => x.FaxNumber)
                    .HasConversion(ValueConverters.PhoneNumberConverter);

                // Shadow properties
                m.Property<string>("Email"); << Used as a FK.
                m.Property<int>("TecTacContactId");
            });
```
If `Profile` related mappings are added **after** those for the `User` entity, EF Core will end up throwing the following exception:
`System.InvalidOperationException: 'The types of the properties specified for the foreign key {'Email'} on entity type 'Profile' do not match the types of the properties in the principal key {'_email'} on entity type 'User'.'`

If `Profile` related mappings are added **before** those for the `User` entity, it works fine...

To me the order should not matter?

Note: If I add a private field called _email and updating the mappings, then order does not matter anymore. It seems to be related to shadow properties?

### Further technical details
EF Core version: 2.1.3
Database Provider: Microsoft.EntityFrameworkCore.SqlServer

### PR DESCRIPTION
**Pull request by smitpatel**
Issue: In this case, the FK property was in shadow state. So when defining the shadow property, we used the principal key's type. Later, HasPrincipalKey changed the principal key properties changing its type but we were losing information about FK properties set explicitly since they were not compatible.
Fix: Preserve old FK properties and reuse them if they were set explicitly and does not have type configured.

Resolves #13300

### CODE REVIEW

**Comment by smitpatel** [comment]
Updated.

File: src/EFCore/Metadata/Builders/ReferenceCollectionBuilderBase.cs [hunk 1]
```diff
@@ -60,11 +61,11 @@ public class ReferenceCollectionBuilderBase : IInfrastructure<IMutableModel>, II
                 PrincipalEntityType = oldBuilder.PrincipalEntityType;
                 DependentEntityType = oldBuilder.DependentEntityType;
                 _foreignKeyProperties = foreignKeySet
-                    ? builder.Metadata.Properties
-                    : oldBuilder._foreignKeyProperties;
+                        ? builder.Metadata.Properties
+                        : DependentEntityType.Builder.GetActualProperties(oldBuilder._foreignKeyProperties, null);
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Builders/ReferenceCollectionBuilderBase.cs]: Extra indent

**Comment by smitpatel** [comment]
Updated.

File: src/EFCore/Metadata/Internal/InternalRelationshipBuilder.cs [hunk 2]
```diff
@@ -1642,6 +1645,12 @@ private bool CanInvert(ConfigurationSource? configurationSource)
                     return false;
                 }
 
+                if (Metadata.GetForeignKeyPropertiesConfigurationSource().Overrides(ConfigurationSource.Explicit)
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/InternalRelationshipBuilder.cs]: DataAnnotation

File: src/EFCore/Metadata/Internal/InternalRelationshipBuilder.cs [hunk 3]
```diff
@@ -1605,15 +1605,18 @@ private bool CanInvert(ConfigurationSource? configurationSource)
...
+                principalEndConfigurationSource: properties == null ? (ConfigurationSource?)null : configurationSource,
+                oldDependentProperties: oldDependentProperties);
         }
 
         private bool CanSetPrincipalKey(
             IReadOnlyList<Property> properties,
             ConfigurationSource? configurationSource,
-            out bool resetDependent)
+            out bool resetDependent,
+            out IReadOnlyList<Property> oldDependentProperties)
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/InternalRelationshipBuilder.cs]: Something like `oldNameDependentProperties`

File: src/EFCore/Metadata/Internal/InternalRelationshipBuilder.cs [hunk 4]
```diff
@@ -1783,7 +1794,8 @@ private bool CanInvert(ConfigurationSource? configurationSource)
             bool removeCurrent,
             bool oldRelationshipInverted,
             ConfigurationSource? principalEndConfigurationSource,
-            ConfigurationSource? configurationSource)
+            ConfigurationSource? configurationSource,
+            [CanBeNull] IReadOnlyList<Property> oldDependentProperties)
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/InternalRelationshipBuilder.cs]: Move this to go after `dependentProperties`

File: src/EFCore/Metadata/Internal/InternalRelationshipBuilder.cs [hunk 5]
```diff
@@ -2482,9 +2498,21 @@ private bool CanInvert(ConfigurationSource? configurationSource)
...
             {
-                var principalKey = principalProperties == null
-                    ? null
-                    : principalEntityType.RootType().Builder.HasKey(principalProperties, configurationSource).Metadata;
+                Key principalKey = null;
+                if (principalProperties != null)
+                {
+                    principalKey = principalEntityType.RootType().Builder.HasKey(principalProperties, configurationSource).Metadata;
+
+                    if (dependentProperties == null
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/InternalRelationshipBuilder.cs]: Change this into an assert inside the `if`, there shouldn't be any cases when both are specified.

Comment (smitpatel) [reply src/EFCore/Metadata/Internal/InternalRelationshipBuilder.cs]: Added ` Debug.Assert(dependentProperties == null || oldNameDependentProperties == null); ` before this line.

File: test/EFCore.InMemory.FunctionalTests/Query/QueryBugsInMemoryTest.cs [hunk 6]
```diff
@@ -863,6 +862,130 @@ public class Entityss
 
         #endregion
 
+        #region Bug13300
+
+        [Fact]
+        public virtual void Explicitly_set_shadow_FK_name_is_preserved_with_HasPrincipalKey()
```
Comment (AndriySvyryd) [inline test/EFCore.InMemory.FunctionalTests/Query/QueryBugsInMemoryTest.cs]: Move this and `HasForeignKey_infers_type_for_shadow_property_when_not_specified` to a new file - `EFCore.Tests\ModelBuilding\ModelBuilderTest.Other.cs`. Tests from `UnmappedTypesTest` can also go there.

File: src/EFCore/Metadata/Internal/InternalRelationshipBuilder.cs (hunk 5 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/InternalRelationshipBuilder.cs]: You should be able to remove this `dependentProperties == null` now

File: test/EFCore.InMemory.FunctionalTests/Query/QueryBugsInMemoryTest.cs (hunk 6 above)
Comment (AndriySvyryd) [reply test/EFCore.InMemory.FunctionalTests/Query/QueryBugsInMemoryTest.cs]: You don't need to use the `TestModelBuilder` in these tests

**Pull request merged**